vms/evm/warp/backend.go

```diff
@@ -0,0 +1,230 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package warp
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ava-labs/libevm/metrics"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p/acp118"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/vms/evm/uptimetracker"
+	"github.com/ava-labs/avalanchego/vms/evm/warp/message"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp/payload"
+)
+
+const (
+	ParseErrCode = iota + 1
+	VerifyErrCode
+)
+
+var _ acp118.Verifier = (*acp118Adapter)(nil)
+
+// BlockStore provides access to accepted blocks.
+type BlockStore interface {
+	GetBlock(ctx context.Context, blockID ids.ID) error
+}
+
+// DB stores and retrieves warp messages from the underlying database.
+type DB struct {
+	db database.Database
+}
+
+// NewDB creates a new warp message database.
+func NewDB(db database.Database) *DB {
+	return &DB{
+		db: db,
+	}
+}
+
+// Add stores a warp message in the database and cache.
+func (d *DB) Add(unsignedMsg *warp.UnsignedMessage) error {
+	msgID := unsignedMsg.ID()
+
+	// In the case when a node restarts, and possibly changes its bls key, the cache gets emptied but the database does not.
+	// So to avoid having incorrect signatures saved in the database after a bls key change, we save the full message in the database.
+	// Whereas for the cache, after the node restart, the cache would be emptied so we can directly save the signatures.
+	if err := d.db.Put(msgID[:], unsignedMsg.Bytes()); err != nil {
+		return fmt.Errorf("failed to put warp message in db: %w", err)
+	}
+
+	return nil
+}
+
+// Get retrieves a warp message from the database.
+func (d *DB) Get(msgID ids.ID) (*warp.UnsignedMessage, error) {
+	unsignedMessageBytes, err := d.db.Get(msgID[:])
+	if err != nil {
+		return nil, err
+	}
+
+	unsignedMessage, err := warp.ParseUnsignedMessage(unsignedMessageBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse unsigned message %s: %w", msgID.String(), err)
+	}
+
+	return unsignedMessage, nil
+}
+
+// Verifier implements acp118.Verifier and validates whether a warp message should be signed.
+type Verifier struct {
+	db            *DB
+	blockClient   BlockStore
+	uptimeTracker *uptimetracker.UptimeTracker
+
+	// Metrics
```

Comment:
This comment is self-documenting - we should avoid "what" comments (as that information can be inferred from reading the code) and focus on documenting "why". I would just remove this comment.